### PR TITLE
add muster time windows structure

### PR DESCRIPTION
**Pull request by pfrollo**

### CODE REVIEW

File: .run/Muster Postgres Controller.run.xml [hunk 1]
```diff
@@ -0,0 +1,22 @@
+<component name="ProjectRunConfigurationManager">
+  <configuration default="false" name="Muster Postgres Controller" type="mocha-javascript-test-runner" nameIsGenerated="true">
```
Comment (richard3d) [inline .run/Muster Postgres Controller.run.xml:2]: What are these files for? Is this part of the attempt to have separate ITs and UTs?

File: packages/server/src/api/muster/muster.postgres.controller.ts [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+import { getEdipi, getRosters, getUnitIds } from '../../util/roster-utils';
+import { getMusterTimeView } from './muster.timeview';
+
+
+class MusterPostgresCtr {
+
+  /**
+   * Provides Muster Compliance details for all service members for the given unit id(s)
+   */
+  async getMusterRoster(req: ApiRequest<OrgRoleParams, null, GetMusterRosterQuery>, res: Response<Paginated<Partial<MusterCompliance>>>) {
```
Comment (richard3d) [inline packages/server/src/api/muster/muster.postgres.controller.ts:18]: This suggestion is wordy but perhaps a more appropriate name is `getUserMusterComplianceForRosters` or something?

File: packages/server/src/mocha-setup.ts [hunk 3]
```diff
@@ -19,15 +19,24 @@ import { Workspace } from './api/workspace/workspace.model';
 import server from './index';
 import { ReportSchema } from './api/report-schema/report-schema.model';
 
+// Set NO_MOCHA_INTEGRATION to true when running test not requiring database or server
```
Comment (richard3d) [inline packages/server/src/mocha-setup.ts:22]: As per our discussion...let's either remove this in the PR and just treat all tests as "integration" tests for now or address the unit and integration test shortfalls now.

**Review by richard3d — CHANGES_REQUESTED** [review]

File: packages/server/package.json [hunk 4]
```diff
@@ -75,6 +75,7 @@
     "form-data": "^2.3.3",
     "http-proxy-middleware": "^0.20.0",
     "json-2-csv": "^3.7.8",
+    "later": "^1.2.0",
```
[the remark this answers is not in the document]
Comment (pfrollo) [reply packages/server/package.json]: I am getting this: TS7016: Could not find a declaration file for module '@breejs/later'. '/Users/work/dev/workspace-dds/covid19-reports/node_modules/@breejs/later/lib/index.js' implicitly has an 'any' type.   Try `npm i --save-dev @types/breejs__later` if it exists or add a new declaration (.d.ts) file containing `declare module '@breejs/later';`

And I can't find the right declaration. The suggested one doesn't work. I will suppress this issue for now, test it and see what happens.

File: packages/server/src/api/muster/muster.compliance.ts [hunk 5]
```diff
@@ -0,0 +1,49 @@
+import { Roster } from '../roster/roster.model';
+import { FilteredObservation } from '../observation/observation.model';
+import { MusterUnitTimeView } from './muster.timeview';
+
+export function toMusterCompliance(rosters: Roster[], observations: FilteredObservation[], musterTimeView: MusterUnitTimeView)
```
[the remark this answers is not in the document]
Comment (pfrollo) [reply packages/server/src/api/muster/muster.compliance.ts]: fixed

File: packages/server/src/api/muster/muster.days.ts [hunk 6]
```diff
@@ -0,0 +1,32 @@
+import { DaysOfTheWeek } from '../../util/util';
+
```
[the remark this answers is not in the document]
Comment (pfrollo) [reply packages/server/src/api/muster/muster.days.ts]: moved

File: packages/server/src/api/observation/observation.model.ts [hunk 7]
```diff
@@ -45,3 +46,17 @@ export class Observation extends BaseEntity {
   @Column({ length: 100, nullable: true})
   reportingGroup?: string;
 }
+
+export function getObservations(edipis: string[], fromDate: moment.Moment, toDate: moment.Moment) : Promise<FilteredObservation[]> {
```
[the remark this answers is not in the document]
Comment (pfrollo) [reply packages/server/src/api/observation/observation.model.ts]: There is one example in this repo done like this. But more common is for this code to exist in a utils module. I will move it to a new utils module as that seems to be more common in this repo.

File: .run/Muster Postgres Controller.run.xml (hunk 1 above)
Comment (pfrollo) [reply .run/Muster Postgres Controller.run.xml:2]: No, these are IntelliJ specific files, just like the `.vscode` folder you can see in this repo. These are stored test run configuration allowing IntelliJ users to conveniently
Screen Shot 2021-07-28 at 12 23 12 PM
 click on a test and run it.

File: packages/server/src/mocha-setup.ts (hunk 3 above)
Comment (pfrollo) [reply packages/server/src/mocha-setup.ts:22]: Would you be OK to wait until this work gets prioritized? https://app.asana.com/0/1188940305683068/1200662026396299/f

File: packages/server/package.json (hunk 4 above)
[the remark this answers is not in the document]
Comment (larskendall) [reply packages/server/package.json]: If there are no types you should be able to add this line to `declarations.d.ts`. It fixed the issue when I tried it.

```typescript
declare module '@breejs/later';
```

[the remark this answers is not in the document]
Comment (pfrollo) [reply packages/server/package.json]: sweet! done

File: packages/server/src/api/muster/muster.timeview.ts [hunk 8]
```diff
@@ -0,0 +1,208 @@
+import { MusterConfiguration, MusterConfWithDateArray } from '@covid19-reports/shared';
+import { In } from 'typeorm';
+import moment, { Moment } from 'moment-timezone';
+import later from 'later';
+import { Org } from '../org/org.model';
+import { Unit } from '../unit/unit.model';
+import { daysToString } from './muster.days';
+
```
[the remark this answers is not in the document]
Comment (pfrollo) [reply packages/server/src/api/muster/muster.timeview.ts]: added

File: packages/server/src/api/observation/observation.model.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (pfrollo) [reply packages/server/src/api/observation/observation.model.ts]: moved

File: packages/server/src/api/muster/muster.postgres.controller.ts (hunk 2 above)
Comment (pfrollo) [reply packages/server/src/api/muster/muster.postgres.controller.ts]: renamed

File: packages/server/package.json (hunk 4 above)
[the remark this answers is not in the document]
Comment (pfrollo) [reply packages/server/package.json]: I added the line to declarations.d.ts but the server would not start: 

[server] /Users/work/dev/workspace-dds/covid19-reports/node_modules/ts-node/src/index.ts:513
[server]     return new TSError(diagnosticText, diagnosticCodes)
[server]            ^
[server] TSError: ⨯ Unable to compile TypeScript:
[server] src/api/muster/mustering.opportunities.ts(4,19): error TS7016: Could not find a declaration file for module '@breejs/later'. '/Users/work/dev/workspace-dds/covid19-reports/node_modules/@breejs/later/lib/index.js' implicitly has an 'any' type.
[server]   Try `npm i --save-dev @types/breejs__later` if it exists or add a new declaration (.d.ts) file containing `declare module '@breejs/later';`

[the remark this answers is not in the document]
Comment (pfrollo) [reply packages/server/package.json]: I tried to add this to tsconfig.json and moved the file to .types but it did not work:

{
  "extends": "../tsconfig.json",
  "typeRoots": [
    "./node_modules/@types",
    "./types"
  ]
}

**Pull request merged**